src/Compilers/CSharp/Portable/BoundTree/BoundDecisionDag.cs

```diff
@@ -196,6 +194,35 @@ BoundDecisionDagNode makeReplacement(BoundDecisionDagNode dag, Func<BoundDecisio
             }
         }
 
+        /// <summary>
+        /// To facilitate subsumption-checking for list-patterns, some nodes may be inserted into the dag during construction.
+        /// We will remove such nodes immediately after as these are not important in lowering or any other analysis.
+        /// </summary>
+        public BoundDecisionDag RemoveCompilerGeneratedNodes()
+        {
+            return Rewrite(static (node, replacement) =>
+            {
+                if (node.WasCompilerGenerated)
+                {
+                    switch (node)
+                    {
+                        case BoundEvaluationDecisionDagNode evalNode:
+                            Debug.Assert(evalNode.Evaluation is BoundDagAssignmentEvaluation);
+                            return replacement[evalNode.Next];
+
+                        case BoundTestDecisionDagNode testNode:
+                            Debug.Assert(testNode.Test is BoundDagValueTest);
```

Comment:
Consider commenting to explain why we pick the WhenFalse branch.
nit: is there a way to tighten this assertion? (we're dealing with a test on a length-or-count value)